python/pip_install/pip_compile.py

```diff
@@ -0,0 +1,89 @@
+"Set defaults for the pip-compile command to run it under Bazel"
+
+import os
+import sys
+from shutil import copyfile
+
+from piptools.scripts.compile import cli
+
+if len(sys.argv) < 4:
+    print(
+        "Expected at least two arguments: requirements_in requirements_out",
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+requirements_in = sys.argv.pop(1)
+requirements_txt = sys.argv.pop(1)
+update_target_name = sys.argv.pop(1)
+
+# Before loading click, set the locale for its parser.
+# If it leaks through to the system setting, it may fail:
+# RuntimeError: Click will abort further execution because Python 3 was configured to use ASCII
+# as encoding for the environment. Consult https://click.palletsprojects.com/python3/ for
+# mitigation steps.
+os.environ["LC_ALL"] = "C.UTF-8"
+os.environ["LANG"] = "C.UTF-8"
+
+UPDATE = True
+# Detect if we are running under `bazel test`
+if "TEST_TMPDIR" in os.environ:
+    UPDATE = False
+    # pip-compile wants the cache files to be writeable, but if we point
+    # to the real user cache, Bazel sandboxing makes the file read-only
+    # and we fail.
+    # In theory this makes the test more hermetic as well.
+    sys.argv.append("--cache-dir")
+    sys.argv.append(os.environ["TEST_TMPDIR"])
+    # Make a copy for pip-compile to read and mutate
+    requirements_out = os.path.join(
+        os.environ["TEST_TMPDIR"], os.path.basename(requirements_txt) + ".out"
+    )
+    copyfile(requirements_txt, requirements_out)
+
+elif "BUILD_WORKING_DIRECTORY" in os.environ:
+    os.chdir(os.environ['BUILD_WORKING_DIRECTORY'])
+else:
+    print(
+        "Expected to find BUILD_WORKING_DIRECTORY in environment",
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+update_target_pkg = "/".join(requirements_in.split('/')[:-1])
+# $(rootpath) in the workspace root gives ./requirements.in
+if update_target_pkg == ".":
+    update_target_pkg = ""
+update_command = "bazel run //%s:%s" % (update_target_pkg, update_target_name)
+
+os.environ["CUSTOM_COMPILE_COMMAND"] = update_command
+
+sys.argv.append("--generate-hashes")
+sys.argv.append("--output-file")
+sys.argv.append(requirements_txt if UPDATE else requirements_out)
+sys.argv.append(requirements_in)
+
+if UPDATE:
+    print("Updating " + requirements_txt)
+    cli()
+else:
+    # cli will exit(0) on success
+    try:
+        print("Checking " + requirements_txt)
+        cli()
```

Comment:
Apologies if I am misunderstanding, but IIUC this is going to call out to PyPI to check if there are newer versions of packages, making the test non-cacheable. One way around this would be to add the `external` tag so that the test results are never cached.

We've taken a somewhat different approach to the problem-we make a hash of the `requirements.in` file and prepend it to the `requirements.txt` file. Then there's a test checking that the hash is still correct. This is a weaker test-it doesn't tell you when you need to update, it instead only makes sure that people remember to update the `requirements.txt` whenever they edit the `requirements.in`, but it is hermetic.

Anyway, some comments from the peanut gallery...

Reply:
That's a good question, and was my first reaction as well. However pip-compile has a separate flag that expresses the intent of "give me newer versions"
https://github.com/jazzband/pip-tools#updating-requirements
and since it reads both `requirements.in` and `requirements.txt` it's possible for it to use the latter to pin its own "re-locking" semantics to give you the same versions you have already.

Also we've been operating this rule for a few months at one client, and haven't seen a case of CI failing because some package cut a release and the pip-compile test suddenly thinks the lockfile is out-of-date.

So I'm pretty confident we are good.

Reply:
Ah ha, thanks for pointing that out; I played around with it locally and it indeed did not work like I thought it did.